### PR TITLE
Dynamic group by

### PR DESCRIPTION
**Pull request by marioloko**
Added group by by an unlimited number of fields.
To achieve this a `DynamicZip` iterator has been added with the aim to let zip an unknown at compile time number of iterators.

WARNING: In contrast with `zip` method which returns a `tuple`, `DynamicZip` returns a `Vec`, that means that the return object is non-copyable. In order, to be able to implement the solution, the `Copy` constraint in `hash_join::prepare_hashed_relation` and `group_by::groupby`, so calls to this functions can incur in a move.

Also added test to check this new functionality, doing a group-by 6 fields and checking the result is the expected one.

### CODE REVIEW

File: polars/src/utils.rs [hunk 1]
```diff
@@ -62,6 +62,47 @@ pub fn get_iter_capacity<T, I: Iterator<Item = T>>(iter: &I) -> usize {
...
+}
+
+impl<I, T> Iterator for DynamicZip<I>
+where
+    I: Iterator<Item = T>,
+{
+    type Item = Vec<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iterators.iter_mut().map(|iter| iter.next()).collect()
```
Comment (ritchie46) [inline polars/src/utils.rs:85]: Every iteration is a new heap allocation with this iterator. Seems good for a dynamic number of keys, but let's keep the old iterators for up to 5 keys.

**Review by ritchie46 — COMMENTED** [review]
Seems like a good feature. I only think it is needlessly expensive if we have a few keys. So let's take best of both worlds. Use the old zip methods for up to 5 keys and above the `DynamicZip` iterator.

File: polars/src/utils.rs (hunk 1 above)
Comment (marioloko) [reply polars/src/utils.rs:85]: Yes I agree with you, it is better to keep the old way to avoid allocation for a small set of keys, and the new way just for dynamic grouping.

I propose to expand from 5 to 12 using the old way, as 12 is the maximum size for a tuple to implement the Copy trait, as explained in the [docs](https://doc.rust-lang.org/std/primitive.tuple.html#trait-implementations-1), and after that point to implement this solution.

Do you agree with this or do you prefer just 5?

Comment (ritchie46) [reply polars/src/utils.rs:85]: Yes 12 is good. The 5 was just some arbitrary chosen number.

Comment (marioloko) [reply polars/src/utils.rs:85]: I have the solution almost ready I am preparing the test to catch the corner cases.

**Comment by marioloko** [comment]
Updated code to perform:
- Static group-by up to 12 elements (Maximum size for which `tuple` implement Copy in Rust), so no heap location is needed. (Recursive macro used to avoid repeating code).
- Dynamic group-by with 13 elements or more (heap allocations are performed).

Added test for:
- Group by 12 elements (Maximum static group-by).
- Group by 13 elements (Minimum dynamic group-by).

File: polars/src/frame/group_by.rs [hunk 2]
```diff
@@ -163,50 +163,76 @@ impl DataFrame {
...
+                    .iter()
+                    .map(|sk| sk.as_groupable_iter())
+                    .collect::<Result<Vec<_>>>()?
+                    .into_dynamic_zip();
+
+                groupby(iter)
             }
         };
 
+        println!("GROUPS: {:?}", groups);
```
Comment (ritchie46) [inline polars/src/frame/group_by.rs:234]: Lost print

**Review by ritchie46 — CHANGES_REQUESTED** [review]
I only found a lost print statement. If that one is gone, I will merge. And thanks for helping. :)

**Comment by marioloko** [comment]
Done! I removed that! Thank you!
